packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp

```diff
@@ -0,0 +1,200 @@
+  model LanguageClientNameModel {
+    @doc("Pass in true")
+    @clientName("CSName", "csharp")
+    @clientName("GoName", "go")
+    @clientName("JavaName", "java")
+    @clientName("TSName", "javascript")
+    @clientName("python_name", "python")
+    defaultName: boolean;
+  }
+
+  model JsonEncodedNameModel {
+    @doc("Pass in true")
+    @encodedName("application/json", "wireName")
+    defaultName: boolean;
+  }
+
+  model ClientNameModel {
+    @doc("Pass in true")
+    @clientName("clientName")
+    defaultName: boolean;
+  }
+
+  model ClientNameAndJsonEncodedNameModel {
+    @doc("Pass in true")
+    @clientName("clientName")
+    @encodedName("application/json", "wireName")
+    defaultName: boolean;
+  }
+
+  @scenario
+  @scenarioDoc("""
+  Testing that we can project the JSON name on the wire from defaultName -> wireName.
+  Your generated SDK should generate JsonEncodedNameModel with one property `defaultName` with wire name `wireName`.
+
+  Expected request body:
+  ```json
+  {"wireName": true}
+  ```
+  """)
+  @route("/json")
+  @post
+  op json(...JsonEncodedNameModel): NoContentResponse;
+
+  @scenario
+  @scenarioDoc("""
+Testing that we can project the client name in our generated SDKs.
+Your generated SDK should generate ClientNameModel with one property `clientName` with wire name `defaultName`.
+
+Expected request body:
+```json
+{"defaultName": true}
+```
+""")
+  @route("/client")
+  @post
+  op client(...ClientNameModel): NoContentResponse;
+
+  @scenario
+  @scenarioDoc("""
+Testing that we can project the language specific name in our generated SDKs.
+Your generated SDK should generate LanguageClientNameModel with one property with your language specific property name and wire name `defaultName`.
+
+Expected request body:
+```json
+{"defaultName": true}
+```
+""")
+  @route("/language")
+  @post
+  op language(...LanguageClientNameModel): NoContentResponse;
+
+  @scenario
+  @scenarioDoc("""
+Testing that we can project the client name and the wire name.
+Your generated SDK should generate ClientNameAndJsonEncodedNameModel with one property with client name `clientName` and wire name `wireName`.
+
+Expected request body:
+```json
+{"wireName": true}
+```
+""")
+  @route("/json-and-client")
+  @post
+  op jsonAndClient(...ClientNameAndJsonEncodedNameModel): NoContentResponse;
+}
+
+
+@scenario
+@scenarioDoc("""
+Testing that we can project the operation name.
+Your generated SDK should generate an operation called `clientName`.
+
+Expected status code: 204
+""")
+@route("/operation")
+@clientName("clientName")
+@post
+op operation(): NoContentResponse;
+
+@scenario
+@scenarioDoc("""
+Testing that we can project a parameter name.
+Your generated SDK should generate an operation `parameter` with a single parameter called `clientName`.
+
+Expected query parameter: `query-name="true"`
+
+""")
+@route("/parameter")
+@post
+op parameter(
+  @clientName("clientName")
+  @encodedName("application/json", "query-name")
```

Comment:
It is a little confuse to define a query parameter for content-type `application/json`. content-type is for body parameter.

Reply:
Thanks, I also found currently `@encodedName` is not working with the query and header parameter. left a comment here https://github.com/Azure/cadl-ranch/pull/511#discussion_r1495377968

Reply:
updated.